doc/v8boinc.1.md
v8boinc(1) -- User Manual
=========================

## SYNOPSIS

    v8boinc --help
    v8boinc <command>
            [--help|-h]
            [--config=<path>|-h <path>]
            [--minor|-m] [--major|-M] [--hold|-H]
            [--save|-s] [--suppress|-S] [--limit=<value>|-j <value]
            [--trace|-w]
            [--debug|-d] [--time|-t] [--less|-L] [--log=<path>|-l <path>]
            [<args>]

## DESCRIPTION

`v8boinc` is a command line program that lets you develop and deploy parallel
applications for V8-BOINC system.

While V8-BOINC system only supports tasks consisted of one `.js` file and one
input `.json` file, `v8boinc` program enables writing a modular applications
using well-known node module system.

Other than using multiple files to develop application, `v8boinc` includes
functionality to unit test the application in exact same environment that is
used on volunteer hosts.

In this document we use the term **application** to denote scientific
distributed application that will be deployed on V8-BOINC system.

## COMMANDS AND WORKFLOW

Commands are (in order of normal workflow):

* **authenticate MASTER_URL**
* **benchmark**
* **init APPLICATION-NAME**
* **test**
* **estimate**
* **run**

Before starting to use `v8boinc` on a new machine, one needs to `authenticate`
and `benchmark` the machine. Authentication step requires that user has an
account on a V8-BOINC server that has submission privileges. Benchmark will
estimate speed of current machine in order to later estimate work load of each
job submitted.

When starting to develop an application for the first time, you need to chdir
to root of the application (or in an empty dir) and issue `v8boinc init`.
This will create hidden `.v8boinc` dir and `app.json` with some configuration
parameters. This command will also download `jsapp` executable for your
platform and verifiy its digital signature. This is a statically compiled
wrapper around V8 that simply runs the script from working directory. See
`v8boinc-jsapp(1)` for more info.

The development cycle is:

* `v8boinc test` while all unit tests are good.
* `v8boinc estimate` while workload of one task is good
* `v8boinc run`!
* connect all of your devices to V8-BOINC and ask your friends to do the same!

`v8boinc estimate` tries to estimate number of floating points operations
required to finish one task. This number is then sent to BOINC server which
will use it to do the job scheduling.

All commands can be shortened to first leter, eg. `v8boinc r` is equivalent to
`v8boinc run`. `run` command will also repeat `test` and `estimate` commands
and proceed only if those commands returned no errors.

The rest of the document shows directory structure of an application and
afterwards, description of each command with flags is presented.

## APPLICATION ANATOMY

    -- git ignored:
    | .v8boinc/
    | | bin/
    | | | jsapp
    | | test-results/
    | | bundles/
    | node_modules/
    -- not ignored:
    [package.json]
    | app.json
    | main.js
    | lib/
    | | *.js
    | input/
    | | *.in.json
    | output/
    | | *.out.json
    | | *.error
    | test/
    | | *.in.json
    | | *.out.json

This diagram shows directory structure of an application. Ignored `.v8boinc`
directory contains `jsapp` executable that runs unit tests, bundles that are
concating whole application code with required modules to one file and unit
test results from last test run.

`main.js` file must `module.exports` a function called `main`. This is a entry
point to your application. This function will be invoked with input object and
returned value will act as an output object. If you want to have function
`main` in some other file change `main` field in `app.json`.

From your code you can use `require` functionality and require anything that is
pure JavaScript. For example, you can't `require("child_process")` because we
can't run that from browser.

Folder `lib/` is used for your libraries just like in ordinary node packages.
Use `npm install --save [package-name]` when you want to use some external
package that will later be installed easily with `npm install` when positioned
in root directory of your app. 

In `test/` you should put `X.in.json` and `X.out.json` pairs that are matching
inputs and outputs from your application. It is required to have at least one
test in order to submit jobs to remote server. Note that these tests will be
started locally. There is a bug that causes all tests to take at least 2
seconds to execute.

`input/` directory should contain real tasks that you want to process in
distributed manner. These files are in same format as your unit tests. The only
difference is that there can be (should be) large number of these tasks and
they can consume significant processing time to process. Other than that,
everything is the same like with unit tests.

`output/` will contain results from running an application distributed, eg.
after `v8boinc run` finishes. These files will have same base name as input
counterparts but extension will be `.out.json` if everything went well or
`.error` if there was an error while processing a task in the wild.

You may choose to have scripts that will produce all the input files or to git
ignore output files and to process them to some other location after the
computatino is finished. It is expected that number of these files will be
huge.

## v8boinc authenticate SERVER-URL [AUTHENTICATOR-KEY]

This command will create global config file with authenticator and submission
server URL.

In order to get an authenticator, navigate to V8-BOINC aware server and then:

* log in or create an account (link is in upper right corner)
* ask submit permissions from server administrator
* click on your account, then account keys
* use 32 char string as your authenticator-key

Args and flags.

  **SERVER-URL**
      BOINC master url of the V8-BOINC project in use.

  **[AUTHENTICATOR]**
      Authenticator key for user that is using V8-BOINC server. If omitted, key
      will be prompted through tty like a password.

  **--config=PATH** default `~/.v8boinc.conf`
      Makes use of different path for global config file. If different config
      file is in use than all future invocations should have that flag active.

## v8boinc benchmark

Benchmark command will estimate speed of current machine and write it to main
config file (`~/.v8boinc.conf`). This data will later be used to estimate
load required by submitted jobs.

## v8boinc init NAME [DESCRIPTION]

This command will create `.gitignore` file, internally used `.v8boinc` folder
and minimal `app.json` for your scientific application and download and
verify binaries used to test the application locally.

While `.v8boinc` is git ignored, you must call this command each time you
start working on a freshly cloned app.

  **NAME** default is basename of current directory
      Name to write into app.config. It is used only to name some internal
      files and directories.

  **[DESCRIPTION]**
      Describe yourslef what you're doing. It will be written to `app.json`.

## v8boinc test [TEST-FILES]

Test will bundle up all `.js` files in your application root directory and
create bundled `main.js` that contains your whole application.

Code will then be executed against test files from arguments or all tests in
`test/` if no arguments specified.

All test that didn't crashed will have results written to
`.v8boinc/test-results/`.

  **[TEST_FILES]**
      Tests to run. If omitted all tests in `test/` directory will be
      used.

  **-s, --save**
    Without this flag, if there is a `.in.json` test file without matching
    `.out.json` file, that test will be marked as failed.
    With this flag active, produced output will be written as correct test
    result and test will pass.

  **-S, --suppress**
    Suppress outputing failed test outputs to a console. Helps if tests have
    large input/output files. All (not crashed) test results are always
    available in `.v8boinc/test-results/*.out.json`.

  **-j <NUM>, --limit=<NUM>** default $(nproc)
    Run this number of tests in parallel.

  **-H, --hold**
    With every run of tests, build version of your application is increased.
    This causes `.v8boinc/bundles` to pile up with earlier snapshots of your
    application which can consume a disk space. This flag causes version to
    hold still.
    Versions are written in `app.json`.

  **-m, --minor**
    Bump application minor version.

  **-M, --major**
    Bump application major version.


## v8boinc estimate [INPUT-FILE]

This command will create a bundle and run the application on first input file
in `input/` dir or file name in command line argument. Execution will be
stopped when `boinc_fraction_done` is called with argument equal or larger than
0.01.

Estimated processing power will be determined based on above running time and
benchmark results for this machine.

  **[INPUT_FILE]**
      You may choose an input file that will be used to estimate work load. If
      this argument is omitted, first input file from `input/` directory will
      be used.

## v8boinc run

Run will firstly invoke `v8boinc test && v8boinc estimate` with all flags
passed to run invocation. If those subcommands return no errors, submition of
tasks from `input/` to remote server will commence.

After all tasks were submitted, process will wait for all results to return
and write their output to `output/`. If task produced an error, its output
will have `.error` extension and contents will have the error details.

If you hit Ctrl+C (or send SIGINT) to this process while there are pending
results, all submitted jobs will be canceled on server.

If you hit Ctrl+C twice, or kill a process in any other way, you may left
active tasks on server. Try not to do this.

## FLAGS THAT CONTROL OUTPUT

  **-w <path>, --trace=<path>**
    Write all request/response HTTP pairs into a file for debugging purposes.
    Trace file will not contain authenticator key which is sent over wire.

  **-d, --debug**
    Print some more debugging info. They are not much usefull.

  **-t, --time**
    Output timestamp prepended to every line. Useful for long running jobs.

  **-L, --less**
    Don't output `INFO` lines.

  **-l <path>, --log=<path>**
    Additionally, output everything to a log file.

## SEE ALSO

* v8boinc-jsapp(1)
* v8boinc-model(7)



lib/V8Boinc.js
/*jslint indent: 2, plusplus: true*/
"use strict";

var di = require('./DI.js'),
  defaults = di.get('./defaults.js'),
  async = di.get('async'),
  xml = di.get('./XmlHelper.js'),
  InputFile = di.get('./InputFile.js'),
  V8BoincBatch = di.get('./V8BoincBatch.js');

/*

interface:

  .initialize(done)

  .submitJobs(jobs, batchName, onSubmit)

  .getActiveBatches(done)
  .getBatchStatus(batchId, done)
  .getJobDetails(jobName, done)
  .getJobResult(jobName, done);

  .abortBatch(batchId, done)
  .killBatch(batchId, done)

  .getLatestExecutable(platform, done)
  .getCodeSignPublicKey(done)

  .newBatch()
    .addJob(mainJs, inJson, rsc_fpops_est, done)
    .submit(onSubmit)
    .getFractionDone()

 sniffer(isRequest, body)
*/

module.exports = function (options, sniffer, logger) {
  var that = this;

  function sniff(isRequest, body) {
    if (sniffer) {
      sniffer(isRequest, body);
    }
  }

  /*
   * simple xml object will have unique keys but values will be 1 element arrays.
   * strip out arrays
   */
  function simplifyXmlObject(obj) {
    var key;
    for (key in obj) {
      if (obj.hasOwnProperty(key)) {
        obj[key] = obj[key][0];
      }
    }
    return obj;
  }

  function downloadUrl(url, done) {
    sniff(true, options.project + url);
    di.get('request')(
      options.project + url,
      function (err, response, body) {
        sniff(false, body);
        if (err) {
          return done(err);
        }
        done(null, body);
      }
    );
  }

  function downloadJsonUrl(url, done) {
    downloadUrl(url, function (err, body) {
      var parsed;
      if (err) {
        return done(err);
      }
      try {
        parsed = JSON.parse(body);
      } catch (ex) {
        return done(
          "couldn't parse out.json " + ex + "response: "
            + JSON.stringify(response)
        );
      }
      done(null, parsed);
    });
  }

  function doHttp(body, done) {
    sniff(true, body);
    di.get('request').post(
      options.project + defaults.rpc_handler_path,
      { form: {request: body} },
      function (err, response, body) {
        sniff(false, body);
        if (err) {
          return done(err, response);
        }
        xml.fromXml(body, function (err, data) {
          if (err) {
            return done("coludnt parse response xml: "
                        + err + "\n\n, original: " + body);
          }
          if (data.hasOwnProperty('error')) {
            return done(simplifyXmlObject(data.error));
          }
          return done(null, data);
        });
      }
    );
  }

  function buildRequestXml(rootName, payloadKey, payloadValue) {
    var obj = { authenticator : options.authenticator };
    if (payloadKey) {
      obj[payloadKey] = payloadValue;
    }
    return xml.toXmlSynced(obj, rootName);
  }

  /**
   * [ { id: '23731',
   *   name: 'job_2458247547',
   *   canonical_instance_id: '0',
   *   n_outfiles: '1' },
   *   { id: '23732',
   *   name: 'job_9895339508',
   *   canonical_instance_id: '26987',
   *   n_outfiles: '1' },
   *   { id: '23733',
   *   name: 'job_4180160199',
   *   canonical_instance_id: '26988',
   *   n_outfiles: '1' } ]
   */
  this.getBatchStatus = function (batchId, done) {
    doHttp(
      buildRequestXml('query_batch', 'batch_id', batchId),
      function (err, result) {
        if (err) {
          return done(err);
        }
        done(null, result.batch.job.map(simplifyXmlObject), {
          nerror_jobs: result.batch.nerror_jobs[0],
          fraction_done: result.batch.fraction_done[0],
          credit_estimate: result.batch.credit_estimate[0]
        });
      }
    );
  };

  this.getJobDetails = function (jobName, done) {
    doHttp(
      buildRequestXml('query_completed_job', 'job_name', jobName),
      function (err, result) {
        if (err) {
          return done(err);
        }
        done(null, simplifyXmlObject(result.completed_job));
      }
    );
  };

  // this job should be returned from getBatchStatus
  this.getJobResult = function (jobName, done) {
    downloadJsonUrl(
      "/get_output.php?cmd=workunit_file&wu_name="
        + jobName + "&file_num=0&auth_str="
        + options.authenticator,
      done
    );
  };

  this.submitJobs = function (jobs, batch_name, done) {
    function serializeJob(job) {
      var est = job.rsc_fpops_est;
      if (!job.hasOwnProperty('name')) {
        return "job should have name!";
      }
      if (!est || typeof est !== 'number' || est < 100 || est > 1e20) {
        return "job should have numberic rsc_fpops_est property elem [100, 1e20]";
      }
      if (!(job.in_json instanceof InputFile.InputFile)) {
        return "job.in_json should be instance of InputFile";
      }
      if (!(job.main_js instanceof InputFile.InputFile)) {
        return "job.main_js should be instance of InputFile";
      }
      return {
        rsc_fpops_est : est,
        name : job.name,
        input_files : [
          job.main_js.toObject(),
          job.in_json.toObject()
        ]
      };
    }
    doHttp(
      buildRequestXml('submit_batch', 'batch', {
        app_name : options.app_name,
        batch_name : batch_name,
        jobs : jobs.map(serializeJob)
      }),
      done
    );
  };

  /*
  1: in progress
  2: completed (all jobs either succeeded or had fatal errors)
  3: aborted
  4: retired
  */
  this.getActiveBatches = function (done) {
    function active(batch) {
      return batch.state !== '4';
    }
    function prettify(batch) {
      var it,
        nums = ['id', 'njobs', 'fraction_done', 'nerror_jobs',
          'credit_estimate', 'credit_canonical', 'create_time',
          'expire_time', 'est_completion_time', 'completion_time'],
        stateMap = {
          1: 'in_progress',
          2: 'completed',
          3: 'aborted',
          4: 'retired'
        };
      for (it = 0; it < nums.length; ++it) {
        batch[nums[it]] = Number(batch[nums[it]]);
      }
      batch.state = stateMap[batch.state];
      return batch;
    }
    doHttp(
      buildRequestXml('query_batches'),
      function (err, result) {
        if (err) {
          return done(err);
        }
        done(
          null,
          result.batches.batch
            .map(simplifyXmlObject)
            .filter(active)
            .map(prettify)
        );
      }
    );
  };

  this.abortBatch = function (batchId, done) {
    doHttp(
      buildRequestXml('abort_batch', 'batch_id', batchId),
      function (err) {
        if (err) {
          return done(err);
        }
        done(null);
      }
    );
  };

  this.killBatch = function (batchId, done) {
    that.abortBatch(batchId, function (err) {
      if (err) {
        return done(err);
      }
      doHttp(
        buildRequestXml('retire_batch', 'batch_id', batchId),
        function (err) {
          if (err) {
            return done(err);
          }
          done(null);
        }
      );
    });
  };

  this.initialize = function (done) {
    if (options.project.substr(-1) !== '/') {
      return done("end your project url with /");
    }
    doHttp(
      buildRequestXml('submit_batch', 'batch', {
        app_name : options.app_name,
        batch_name : 'test credentials',
        jobs : []
      }),
      function (err) {
        if (err.error_msg === 'trying to submit empty batch') {
          done(null);
        } else {
          done(err.hasOwnProperty('error_msg') ?
               err.error_msg : err);
        }
      }
    );
  };

  this.newBatch = function () {
    return new V8BoincBatch(that, logger);
  };

  this.getCodeSignPublicKey = function (done) {
    doHttp(
      buildRequestXml('get_code_sign_public_key', 'foo', 'bar'),
      function (err, data) {
        if (err) {
          return done(err);
        }
        done(null, data.public_key);
      }
    );
  };

  this.getLatestExecutable = function (platform, done) {
    async.series([
      downloadUrl.bind(null, "/bin/jstest-" + platform + ".sig.base64"),
      downloadUrl.bind(null, "/bin/jstest-" + platform + ".version"),
    ], function (err, results) {
      if (err) {
        return done(err);
      }
      done(null, {
        url : options.project + "bin/jstest-" + platform,
        file_signature : results[0].trim(),
        version : results[1].trim()
      });
    });
  };

  (function () {
    if (typeof options !== 'object') {
      throw "options should be object";
    }

    options.project = options.project || defaults.project;
    options.app_name = options.app_name || defaults.app_name;
  }());

};

